source/val/validation_state.cpp

```diff
@@ -520,17 +520,39 @@ void ValidationState_t::RegisterDebugInstruction(const Instruction* inst) {
 void ValidationState_t::RegisterInstruction(Instruction* inst) {
   if (inst->id()) all_definitions_.insert(std::make_pair(inst->id(), inst));
 
-  // If the instruction is using an OpTypeSampledImage as an operand, it should
-  // be recorded. The validator will ensure that all usages of an
-  // OpTypeSampledImage and its definition are in the same basic block.
+  // Some validation checks are easier by getting all the consumers
   for (uint16_t i = 0; i < inst->operands().size(); ++i) {
     const spv_parsed_operand_t& operand = inst->operand(i);
-    if (SPV_OPERAND_TYPE_ID == operand.type) {
+    if ((SPV_OPERAND_TYPE_ID == operand.type) ||
+        (SPV_OPERAND_TYPE_TYPE_ID == operand.type)) {
       const uint32_t operand_word = inst->word(operand.offset);
       Instruction* operand_inst = FindDef(operand_word);
-      if (operand_inst && SpvOpSampledImage == operand_inst->opcode()) {
+      if (!operand_inst) {
+        continue;
+      }
+
+      // If the instruction is using an OpTypeSampledImage as an operand, it
+      // should be recorded. The validator will ensure that all usages of an
+      // OpTypeSampledImage and its definition are in the same basic block.
+      if ((SPV_OPERAND_TYPE_ID == operand.type) &&
+          (SpvOpSampledImage == operand_inst->opcode())) {
         RegisterSampledImageConsumer(operand_word, inst);
       }
+
+      // In order to track storage classes (not Function) used per execution
+      // model we can't use RegisterExecutionModelLimitation on instructions
+      // like OpTypePointer which are going to be in the pre-function section.
+      // Instead just need to register storage class usage for consumers in a
+      // function block.
+      if (inst->function()) {
+        if (operand_inst->opcode() == SpvOpTypePointer) {
```

Comment:
Is there an instance where tracking variables is not enough? The only place I can think of would be physical storage buffer, but is not covered by any of the checks that were added.

Reply:
So from the grammar file all storage classes for shaders come from `OpTypePointer` (and forward pointers which will be the same) and `OpVariable`

I guess let me add some tests with physical storage buffers to help make sure

Reply:
So (finally) had time to get back into SPIRV-Tools and looking back at this, I guess I don't see why tracking operands pointing to `OpTypePointer`/`OpVariable` since that will show all the storage classes being used (since `OpTypeForwardPointer` will be the same) (realize this wouldn't track `OpGenericCastToPtrExplicit` currently)

As for `PhysicalStorageBuffer`, it will never cross and point to track to other storage classes (to my knowledge, it can't) so not sure how a storage class could get "hidden" from being captured being used in a function (to later be checked for execution model)